File: android/app/src/main/java/np/nexpay/wallet/core/Protocol.kt
package np.nexpay.wallet.core

import org.json.JSONArray
import org.json.JSONObject
import java.math.BigDecimal
import java.math.RoundingMode
import java.security.*
import java.security.spec.X509EncodedKeySpec
import java.util.Base64
import java.util.UUID

object Protocol {
    const val OFFLINE_LIMIT = 500_000L
    const val NOTE_LIFE = 86_400_000L
    const val REDEEM_GRACE = 604_800_000L
    const val SKEW = 300_000L
    fun b64(bytes: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
    fun unb64(s: String): ByteArray {
        require(s.isNotEmpty() && s.length <= 24000 && s.matches(Regex("[A-Za-z0-9_-]+"))) { "Invalid payment encoding" }
        return Base64.getUrlDecoder().decode(s).also { require(b64(it) == s) { "Non-canonical encoding" } }
    }
    fun sha(bytes: ByteArray): String = MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it.toInt() and 255) }
    fun walletId(pub: String) = "pa_" + sha(unb64(pub)).take(32)
    fun publicKey(pub: String): PublicKey = KeyFactory.getInstance("EC").generatePublic(X509EncodedKeySpec(unb64(pub))).also {
        val ec = it as java.security.interfaces.ECPublicKey
        val spec = AlgorithmParameters.getInstance("EC").apply { init(java.security.spec.ECGenParameterSpec("secp256r1")) }.getParameterSpec(java.security.spec.ECParameterSpec::class.java)
        require(ec.params.order == spec.order && ec.params.curve == spec.curve && ec.params.generator == spec.generator && ec.params.cofactor == spec.cofactor) { "Use a P-256 key" }
        require(b64(it.encoded) == pub) { "Non-canonical public key" }
    }
    fun sign(kind: String, data: JSONObject, key: PrivateKey): String {
        val encoded = b64(data.toString().toByteArray(Charsets.UTF_8))
        val sig = Signature.getInstance("SHA256withECDSA").run { initSign(key); update("paila:$kind:v1:$encoded".toByteArray(Charsets.UTF_8)); sign() }
        return "p1.$encoded.${b64(sig)}"
    }
    fun peek(raw: String): JSONObject {
        require(raw.length <= 23000) { "Payment message too large" }
        val pieces = raw.split('.')
        require(pieces.size == 3 && pieces[0] == "p1") { "Not a NexPay payment code" }
        return JSONObject(String(unb64(pieces[1]), Charsets.UTF_8)).also { require(it.getInt("v") == 1) { "Unsupported payment version" } }
    }
    fun verify(kind: String, raw: String, pub: String): JSONObject {
        val data = peek(raw)
        val pieces = raw.split('.')
        val valid = Signature.getInstance("SHA256withECDSA").run {
            initVerify(publicKey(pub)); update("paila:$kind:v1:${pieces[1]}".toByteArray(Charsets.UTF_8)); verify(unb64(pieces[2]))
        }
        require(valid) { "Payment signature is invalid" }
        return data
    }
    fun readReceive(raw: String, now: Long = System.currentTimeMillis()): JSONObject {
        val d = peek(raw)
        val r = verify("receive", raw, d.getString("publicKey"))
        require(walletId(r.getString("publicKey")) == r.getString("walletId")) { "Recipient key mismatch" }
        safeName(r.getString("name"))
        val start = r.getLong("createdAt"); val end = r.getLong("expiresAt")
        require(start <= now + SKEW && end > now && end - start in 1..NOTE_LIFE) { "Receive code expired. Ask for a new one." }
        require(r.getString("requestId").matches(Regex("[A-Za-z0-9_-]{16,80}"))) { "Invalid receive request" }
        return r
    }
    const val MAX_HOPS = 3
    const val QR_HOPS = 1
    data class Payment(val note: JSONObject, val data: JSONObject, val hash: String, val amount: Long, val hop: Int)
    data class ChainLink(val sender: String, val to: String, val amount: Long, val hash: String)

    fun chainDepth(packet: String): Int = try { peek(packet).optInt("hop", 0) } catch (_: Exception) { 0 }

    fun readPayment(raw: String, issuer: String, recipient: String, now: Long = System.currentTimeMillis(), settlement: Boolean = false): Payment {
        val data = peek(raw)
        val hop = data.optInt("hop", 0)
        require(hop in 0..MAX_HOPS) { "Transfer chain is too long or invalid" }
        val chainArr = data.optJSONArray("chain")
        val links = (0 until (chainArr?.length() ?: 0)).map { chainArr!!.getString(it) }
        require(links.size == hop) { "Transfer chain does not match hop count" }
        val note = verify("voucher", if (hop == 0) data.getString("voucher") else peek(links[0]).getString("voucher"), issuer)
        require(note.getString("issuer") == sha(unb64(issuer))) { "Payment belongs to a different server" }
        require(note.getLong("amount") in 1..OFFLINE_LIMIT) { "Offline amount exceeds limit" }
        require(note.getString("owner") == walletId(note.getString("ownerKey"))) { "Note owner mismatch" }
        require(note.getString("id").matches(Regex("[A-Za-z0-9_-]{16,80}"))) { "Invalid note identifier" }
        val ncreated = note.getLong("createdAt"); val nexpires = note.getLong("expiresAt")
        require(nexpires - ncreated == NOTE_LIFE) { "Invalid note lifetime" }
        var parentAmount = note.getLong("amount"); var parentHash = ""; var parentTo = ""
        for (i in 0 until hop) {
            val ld = peek(links[i])
            require(ld.optInt("hop", 0) == i) { "Broken chain continuity" }
            val lf = if (ld.has("fromKey")) ld.getString("fromKey") else if (i == 0) note.getString("ownerKey") else error("Chain link is missing its sender key")
            publicKey(lf)
            require(ld.getString("requestId").matches(Regex("[A-Za-z0-9_-]{16,80}"))) { "Invalid chain request" }
            if (i == 0) require(walletId(lf) == note.getString("owner")) { "Only the note owner can start a transfer" }
            else { require(walletId(lf) == parentTo) { "Chain sender does not follow the previous hop" }; require(ld.getString("prev") == parentHash) { "Chain link does not reference the previous transfer" } }
            verify("payment", links[i], lf)
            val lto = ld.getString("to"); require(lto != walletId(lf)) { "Choose another wallet" }
            val lamt = ld.getLong("amountMinor"); require(lamt in 1..parentAmount) { "Chain amount exceeds what was received" }
            val lct = ld.getLong("createdAt"); require(lct in (ncreated - SKEW)..nexpires) { "Chain transfer outside note lifetime" }
            val lh = sha(canonical(ld).toByteArray(Charsets.UTF_8))
            parentHash = lh; parentAmount = lamt; parentTo = lto
        }
        val senderKey = if (data.has("fromKey")) data.getString("fromKey") else note.getString("ownerKey")
        publicKey(senderKey)
        if (hop == 0) require(walletId(senderKey) == note.getString("owner")) { "Only the note owner can start a transfer" }
        else { require(walletId(senderKey) == parentTo) { "Transfer sender does not follow the chain" }; require(data.getString("prev") == parentHash) { "Transfer does not reference the previous transfer" } }
        verify("payment", raw, senderKey)
        require(data.getString("to") == recipient) { "Payment is addressed to a different wallet" }
        if (hop == 0) require(recipient != note.getString("owner")) { "Choose another wallet" } else require(recipient != walletId(senderKey)) { "Choose another wallet" }
        val paid = if (data.has("amountMinor")) data.getLong("amountMinor") else parentAmount
        require(paid in 1..parentAmount) { "Payment exceeds the available value" }
        val time = data.getLong("createdAt")
        val earliest = if (hop == 0) ncreated else ncreated
        require(time in (earliest - SKEW)..nexpires && time <= now + SKEW) { "Offline note expired or invalid device clock" }
        require(now < nexpires + if (settlement) REDEEM_GRACE else 0) { "Offline note expired" }
        require(data.getString("requestId").matches(Regex("[A-Za-z0-9_-]{16,80}"))) { "Invalid payment identifier" }
        return Payment(note, data, sha(canonical(data).toByteArray(Charsets.UTF_8)), paid, hop)
    }
    fun canonical(value: Any?): String = when (value) {
        null, JSONObject.NULL -> "null"
        is JSONObject -> value.keys().asSequence().toList().sorted().joinToString(",", "{", "}") { JSONObject.quote(it) + ":" + canonical(value.get(it)) }
        is JSONArray -> (0 until value.length()).joinToString(",", "[", "]") { canonical(value.get(it)) }
        is String -> JSONObject.quote(value)
        is Boolean, is Number -> value.toString()
        else -> error("Unsupported value")
    }
    fun paisa(input: String, max: Long = 10_000_000): Long {
        require(input.trim().matches(Regex("[0-9]{1,8}(\\.[0-9]{1,2})?"))) { "Enter a positive rupee amount with up to two decimals" }
        val result = BigDecimal(input.trim()).setScale(2, RoundingMode.UNNECESSARY).movePointRight(2).longValueExact()
        require(result in 1..max) { "Amount is outside the allowed limit" }
        return result
    }
    fun safeName(name: String): String = name.trim().also {
        require(it.length in 1..48 && !Regex("[\\p{Cc}\\u202A-\\u202E\\u2066-\\u2069]").containsMatchIn(it)) { "Enter a name (1–48 characters, no control characters)" }
    }
    fun newId() = UUID.randomUUID().toString()
    fun obj(vararg pairs: Pair<String, Any?>) = JSONObject().also { j -> pairs.forEach { (k, v) -> j.put(k, v) } }
}

File: server/src/protocol.mjs
import { createHash, createPublicKey, sign, verify, randomUUID, generateKeyPairSync } from 'node:crypto';
export const LIMITS = Object.freeze({ requestBytes:32768, grant:500000, offline:500000, online:10000000, noteLife:86400000, redeemGrace:604800000, clockSkew:300000, maxHops:3, chainBytes:48000, qrHops:1 });
export class Fault extends Error { constructor(code, message, status=400) { super(message); this.code=code; this.status=status; } }
export const fail=(ok,code,message,status=400)=>{if(!ok)throw new Fault(code,message,status);};
export const hash=s=>createHash('sha256').update(s).digest('hex');
export const b64=b=>Buffer.from(b).toString('base64url');
export function unb64(s) {
  fail(typeof s==='string' && s.length>0 && s.length<=48000 && /^[A-Za-z0-9_-]+$/.test(s),'BAD_ENCODING','Invalid base64url.');
  const b=Buffer.from(s,'base64url'); fail(b64(b)===s,'BAD_ENCODING','Non-canonical base64url.'); return b;
}
export function publicKey(encoded) {
  try {
    const der=unb64(encoded); const key=createPublicKey({key:der,format:'der',type:'spki'});
    fail(key.asymmetricKeyType==='ec' && key.asymmetricKeyDetails?.namedCurve==='prime256v1','BAD_KEY','Use an ECDSA P-256 public key.');
    fail(b64(key.export({type:'spki',format:'der'}))===encoded,'BAD_KEY','Use canonical SPKI key encoding.');
    return key;
  } catch(e) { if(e instanceof Fault)throw e; throw new Fault('BAD_KEY','Invalid public key.'); }
}
export const exportKey=key=>b64(key.export({type:'spki',format:'der'}));
export const walletId=pub=>'pa_'+hash(unb64(pub)).slice(0,32);
export const newKey=()=>generateKeyPairSync('ec',{namedCurve:'prime256v1'});
export function pack(kind, data, privateKey) {
  const encoded=b64(JSON.stringify(data));
  const signature=sign('sha256',Buffer.from(`paila:${kind}:v1:${encoded}`),privateKey);
  return `p1.${encoded}.${b64(signature)}`;
}
export function unpack(envelope) {
  fail(typeof envelope==='string' && envelope.length<=65536,'BAD_ENVELOPE','Payment message is too large or invalid.');
  const pieces=envelope.split('.'); fail(pieces.length===3 && pieces[0]==='p1','BAD_ENVELOPE','Invalid signed message.');
  const raw=unb64(pieces[1]); const signature=unb64(pieces[2]);
  fail(signature.length>=8 && signature.length<=72,'BAD_SIGNATURE','Invalid ECDSA signature size.',401);
  let data; try{data=JSON.parse(new TextDecoder('utf-8',{fatal:true}).decode(raw));}catch{throw new Fault('BAD_ENVELOPE','Invalid message JSON.');}
  fail(data && typeof data==='object' && !Array.isArray(data) && data.v===1,'BAD_VERSION','Unsupported protocol version.');
  return {data,encoded:pieces[1],signature};
}
export function open(kind,envelope,key) {
  const p=unpack(envelope);
  let valid=false;try{valid=verify('sha256',Buffer.from(`paila:${kind}:v1:${p.encoded}`),typeof key==='string'?publicKey(key):key,p.signature);}catch{}
  fail(valid,'BAD_SIGNATURE','Signature verification failed.',401); return p.data;
}
export function text(value, label, max=120) {
  fail(typeof value==='string' && value.trim().length>0 && value.length<=max && !/[\u0000-\u001F\u007F-\u009F\u202A-\u202E\u2066-\u2069]/u.test(value),'BAD_INPUT',`${label} is invalid.`);return value.trim();
}
export function amount(n,max=LIMITS.online) {fail(Number.isSafeInteger(n)&&n>0&&n<=max,'BAD_AMOUNT','Amount must be positive integer paisa within the limit.');return n;}
export function fresh(ts, now) {fail(Number.isSafeInteger(ts)&&Math.abs(now-ts)<=LIMITS.clockSkew,'STALE_REQUEST','Check your device clock and retry.',401);}
export function operationId(s){fail(typeof s==='string'&&/^[A-Za-z0-9_-]{16,80}$/.test(s),'BAD_OPERATION_ID','Invalid operation identifier.');return s;}
export const canonical=o=>o===null||typeof o!=='object'?JSON.stringify(o):Array.isArray(o)?`[${o.map(canonical).join(',')}]`:`{${Object.keys(o).sort().map(k=>`${JSON.stringify(k)}:${canonical(o[k])}`).join(',')}}`;
export function request(key,wallet,op,data={},id=randomUUID(),now=Date.now()) {return pack('request',{v:1,walletId:wallet,op,opId:id,ts:now,data},key);}
export function receiveRequest(key,pub,name,now=Date.now()) {return pack('receive',{v:1,walletId:walletId(pub),publicKey:pub,name,requestId:randomUUID(),createdAt:now,expiresAt:now+86400000},key);}
export function readReceive(raw,now=Date.now()) {
  const p=unpack(raw).data; publicKey(p.publicKey);const r=open('receive',raw,p.publicKey);
  fail(walletId(r.publicKey)===r.walletId,'BAD_RECIPIENT','Recipient key mismatch.');text(r.name,'Name',48);operationId(r.requestId);
  fail(Number.isSafeInteger(r.createdAt)&&Number.isSafeInteger(r.expiresAt)&&r.createdAt<=now+LIMITS.clockSkew&&r.expiresAt>now&&r.expiresAt-r.createdAt<=86400000,'EXPIRED_REQUEST','Ask for a fresh receive code.');return r;
}
export function readPayment(raw,issuerKey,recipient,now=Date.now(),settlement=false) {
  const outer=unpack(raw).data;
  const hop=outer.hop===undefined?0:outer.hop;
  fail(Number.isSafeInteger(hop)&&hop>=0&&hop<=LIMITS.maxHops,'BAD_HOP','Transfer chain is too long or invalid.');
  const links=outer.chain===undefined?[]:outer.chain;
  fail(Array.isArray(links)&&links.length===hop,'BAD_CHAIN','Transfer chain does not match hop count.');
  const noteCert=hop===0?outer.voucher:unpack(links[0]).data.voucher;
  const note=open('voucher',noteCert,issuerKey);
  const issuer=hash(typeof issuerKey==='string'?unb64(issuerKey):issuerKey.export({type:'spki',format:'der'}));
  fail(note.issuer===issuer,'WRONG_ISSUER','This note belongs to another server.');
  amount(note.amount,LIMITS.offline);publicKey(note.ownerKey);operationId(note.id);
  fail(walletId(note.ownerKey)===note.owner,'BAD_OWNER','Note owner does not match key.');
  fail(Number.isSafeInteger(note.createdAt)&&Number.isSafeInteger(note.expiresAt)&&note.expiresAt-note.createdAt===LIMITS.noteLife,'BAD_NOTE','Invalid note lifetime.');
  const ancestors=[];let parentHash=null,parentAmount=note.amount,parentTo=null;
  for(let i=0;i<hop;i++){
    fail(typeof links[i]==='string'&&links[i].length<=LIMITS.chainBytes,'BAD_CHAIN','Broken chain link.');
    const ld=unpack(links[i]).data;
    fail((ld.hop===undefined?0:ld.hop)===i,'BAD_CHAIN','Broken chain continuity.');
    const lfrom=typeof ld.fromKey==='string'?ld.fromKey:(i===0?note.ownerKey:null);
    fail(typeof lfrom==='string','BAD_CHAIN','Chain link is missing its sender key.');
    publicKey(lfrom);operationId(ld.requestId);
    if(i===0)fail(walletId(lfrom)===note.owner,'WRONG_SENDER','Only the note owner can start a transfer.');
    else{fail(walletId(lfrom)===parentTo,'BAD_CHAIN','Chain sender does not follow the previous hop.');fail(ld.prev===parentHash,'BAD_CHAIN','Chain link does not reference the previous transfer.');}
    open('payment',links[i],lfrom);
    const lto=text(ld.to,'Recipient',128);
    fail(lto!==walletId(lfrom),'SELF_PAYMENT','Choose another wallet.');
    const lamt=amount(ld.amountMinor,parentAmount);
    fail(Number.isSafeInteger(ld.createdAt)&&ld.createdAt>=note.createdAt-LIMITS.clockSkew&&ld.createdAt<=note.expiresAt,'BAD_TIME','Invalid chain time.');
    const lhash=hash(canonical(ld));
    ancestors.push({sender:walletId(lfrom),to:lto,amount:lamt,hash:lhash,createdAt:ld.createdAt});
    parentHash=lhash;parentAmount=lamt;parentTo=lto;
  }
  const senderKey=typeof outer.fromKey==='string'?outer.fromKey:note.ownerKey;
  if(hop===0)fail(walletId(senderKey)===note.owner,'WRONG_SENDER','Only the note owner can start a transfer.');
  else fail(walletId(senderKey)===parentTo,'BAD_CHAIN','Transfer sender does not follow the chain.');
  const payment=open('payment',raw,senderKey);operationId(payment.requestId);
  fail(payment.to===text(payment.to,'Recipient',128),'BAD_INPUT','Recipient is invalid.');
  if(hop===0)fail(payment.to!==note.owner,'SELF_PAYMENT','Choose another wallet.');
  else fail(payment.to!==walletId(senderKey),'SELF_PAYMENT','Choose another wallet.');
  fail(payment.to===recipient,'WRONG_RECIPIENT','This payment is not addressed to your wallet.',403);
  if(hop>0)fail(payment.prev===parentHash,'BAD_CHAIN','Transfer does not reference the previous transfer.');
  const cap=hop===0?note.amount:parentAmount;
  const amountMinor=payment.amountMinor===undefined?cap:amount(payment.amountMinor,cap);
  const earliest=hop===0?note.createdAt:ancestors[hop-1].createdAt;
  fail(Number.isSafeInteger(payment.createdAt)&&payment.createdAt>=earliest-LIMITS.clockSkew&&payment.createdAt<=note.expiresAt&&payment.createdAt<=now+LIMITS.clockSkew,'BAD_TIME','Invalid payment time.');
  fail(now<note.expiresAt+(settlement?LIMITS.redeemGrace:0),'EXPIRED_NOTE',settlement?'Redemption window has closed.':'This offline note has expired.');
  return {note,noteCert,payment,paymentHash:hash(canonical(payment)),amountMinor,hop,ancestors,senderId:walletId(senderKey)};
}
